app/src/main/java/com/github/polypoly/app/network/RemoteDB.kt

```diff
@@ -0,0 +1,131 @@
+            } else {
+                future.complete(userRef.getValue<User>())
+            }
+        }.addOnFailureListener(future::completeExceptionally)
+
+        return future
+    }
+
+    override fun getAllUsers(): CompletableFuture<List<User>> {
+        val usersPromise = CompletableFuture<List<User>>()
+
+        usersRootRef.get().addOnSuccessListener{ usersRef ->
+            val users = ArrayList<User>()
+            for (child in usersRef.children) {
+                users.add(child.getValue<User>()!!)
+            }
+            usersPromise.complete(users)
+        }.addOnFailureListener(usersPromise::completeExceptionally)
+
+        return usersPromise
+    }
+
+    override fun getAllUsersIds(): CompletableFuture<List<Long>> {
+        val usersIdsPromise = CompletableFuture<List<Long>>()
+
+        usersRootRef.get().addOnSuccessListener{ usersRef ->
+            val ids = ArrayList<Long>()
+            for (child in usersRef.children) {
+                ids.add(child.key?.toLong() ?: -1)
+            }
+            usersIdsPromise.complete(ids)
+        }.addOnFailureListener(usersIdsPromise::completeExceptionally)
+
+        return usersIdsPromise
+    }
+
+    override fun registerUser(user: User): CompletableFuture<Boolean> {
+        return getAllUsersIds().thenCompose { allIds ->
+            val registerPromise = CompletableFuture<Boolean>()
+
+            if (allIds.contains(user.id)) {
+                registerPromise.completeExceptionally(
+                    IllegalAccessError("You cannot register two times a single account"))
+            } else {
+                usersRootRef.child(user.id.toString()).setValue(user).addOnSuccessListener {
+                    registerPromise.complete(true)
+                }.addOnFailureListener(registerPromise::completeExceptionally)
+            }
+
+            registerPromise
+        }
+    }
+
+    override fun updateUser(user: User): CompletableFuture<Boolean> {
+        return getAllUsersIds().thenCompose { allIds ->
+            val registerPromise = CompletableFuture<Boolean>()
+
+            if (!allIds.contains(user.id)) {
+                registerPromise.completeExceptionally(
+                    IllegalAccessError("Tries to update a user not registered yet"))
+            } else {
+                usersRootRef.updateChildren(mapOf(
+                    user.id.toString() to user
+                )).addOnSuccessListener {
+                    registerPromise.complete(true)
+                }.addOnFailureListener(registerPromise::completeExceptionally)
+            }
+
+            registerPromise
+        }
+    }
+
+    fun getUnderlyingDB(): FirebaseDatabase{
+        return db!!
+    }
+
+    companion object InvalidRemoteDB: RemoteDB(null, "") {
+        override fun getUserWithId(userId: Long): CompletableFuture<User> {
+            throw IllegalAccessError("This RemoteDB is invalid")
```

Comment:
This is not the right kind of exception to throw. `Error`s should basically never be thrown by your own code, only by the runtime, they indicate something is very very wrong. In this case `IllegalAccessError` is intended to mean "you're trying to access a field you don't have access to because you're loading incompatible JARs".